tests/ssg_test_suite/test_env.py

```diff
@@ -2,81 +2,188 @@
+        command_base.extend(
+            ['xccdf', 'eval'])
+        return command_base
+
+    def scan(self, args, verbose_path):
+        if self.scanning_mode == "online":
+            return self.online_scan(args, verbose_path)
+        elif self.scanning_mode == "offline":
+            return self.offline_scan(args, verbose_path)
+        else:
+            msg = "Invalid scanning mode {mode}".format(mode=self.scanning_mode)
+            raise KeyError(msg)
+
+    def online_scan(self, args, verbose_path):
+        command_list = self._oscap_ssh_base_arguments() + args
+
+        env = dict(SSH_ADDITIONAL_OPTIONS=" ".join(common.IGNORE_KNOWN_HOSTS_OPTIONS))
+        env.update(os.environ)
+
+        return common.run_cmd_local(command_list, verbose_path, env=env)
+
+    def offline_scan(self, args, verbose_path):
+        raise NotImplementedError()
 
 
 class VMTestEnv(TestEnv):
     name = "libvirt-based"
 
-    def __init__(self, hypervisor, domain_name):
-        super(VMTestEnv, self).__init__()
+    def __init__(self, mode, hypervisor, domain_name):
+        super(VMTestEnv, self).__init__(mode)
+        self.domain = None
 
         self.hypervisor = hypervisor
         self.domain_name = domain_name
         self.snapshot_stack = None
 
+        self._origin = None
+
     def start(self):
-        dom = ssg_test_suite.virt.connect_domain(
+        self.domain = ssg_test_suite.virt.connect_domain(
             self.hypervisor, self.domain_name)
-        self.snapshot_stack = SnapshotStack(dom)
+        self.snapshot_stack = SnapshotStack(self.domain)
 
-        ssg_test_suite.virt.start_domain(dom)
-        self.domain_ip = ssg_test_suite.virt.determine_ip(dom)
+        ssg_test_suite.virt.start_domain(self.domain)
+        self.domain_ip = ssg_test_suite.virt.determine_ip(self.domain)
 
-    def _get_snapshot(self, snapshot_name):
-        return self.snapshot_stack.create(snapshot_name)
+        self._origin = self._save_state("origin")
 
-    def _revert_snapshot(self, snapshot):
+    def finalize(self):
+        self._delete_saved_state(self._origin)
+        # self.domain.shutdown()
+        # logging.debug('Shut the domain off')
+
+    def reset_state_to(self, state_name, new_running_state_name):
+        last_snapshot_name = self.snapshot_stack.snapshot_stack[-1].getName()
+        assert last_snapshot_name == state_name, (
+            "You can only revert to the last snapshot, which is {0}, not {1}"
+            .format(last_snapshot_name, state_name))
+        state = self.snapshot_stack.revert(delete=False)
+        return state
+
+    def discard_running_state(self, state_handle):
+        pass
+
+    def _save_state(self, state_name):
+        super(VMTestEnv, self)._save_state(state_name)
+        state = self.snapshot_stack.create(state_name)
+        return state
+
+    def _delete_saved_state(self, snapshot):
         self.snapshot_stack.revert()
 
+    def _oscap_docker_base_arguments(self):
```

Comment:
I think this function has a confusing name because it is using `oscap-vm` inside, not oscap-docker.